components/salsa-2022/src/interned.rs

```diff
@@ -83,23 +83,39 @@ where
             self.reset_at,
         );
 
+        // Optimisation to only get read lock on the map if the data has already
+        // been interned.
         if let Some(id) = self.key_map.get(&data) {
             return *id;
         }
 
-        loop {
-            let next_id = self.counter.fetch_add(1);
-            let next_id = Id::from_id(crate::id::Id::from_u32(next_id));
-            match self.value_map.entry(next_id) {
-                // If we already have an entry with this id...
-                dashmap::mapref::entry::Entry::Occupied(_) => continue,
-
-                // Otherwise...
-                dashmap::mapref::entry::Entry::Vacant(entry) => {
-                    self.key_map.insert(data.clone(), next_id);
-                    entry.insert(Box::new(data));
-                    return next_id;
-                }
+        let next_id = self.counter.fetch_add(1);
```

Comment:
So actually I was a bit nervous about the `fetch_add` to begin with. It seems like this could cause "gaps" in the interned values. I guess that might be ok. 

Overall this code feels too complex. I'm pondering a few options.

One of them is to just add a "modification lock" that we have to acquire first and foremost. i.e., something like

```rust
let modification_lock = self.modification_lock.lock();
```

that would mean you can only intern (or delete) one thing at a time, but it would remove the concern about deadlock.

Reply:
Actually, wait. Does `delete_index` really have to be *atomic*? Can't we do something like this?

```rust
let key = self.value_map.remove(id).unwrap();
self.key_map.remove(&key);
self.deleted_entries.push(key);
```

After all, the function invariant specifies that there can be no "This should only be used when you are certain that the given `id` has not (and will not) be used in the current revision." OK, well, that's not strong enough, but I think we could strengthen the invariant to also say this should only be used when...

* you are certain that the given `id` has not been and will not be used in the current revision
* you are certain that the interned key corresponding to `id` will not be interned in this revision

The one place we call this is when deleting a tracked struct, and we know that the second condition holds then.

Reply:
That would then allow us to go back to your original formulation where we hold the lock on key-map whilst inserting into value-map.

Reply:
Fair enough, I've done exactly that